compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -165,13 +165,16 @@ pub enum BorrowKind {
     /// Data must be immutable and is aliasable.
     Shared,
 
-    /// The immediately borrowed place must be immutable, but projections from
-    /// it don't need to be. For example, a shallow borrow of `a.b` doesn't
-    /// conflict with a mutable borrow of `a.b.c`.
-    ///
-    /// This is used when lowering matches: when matching on a place we want to
-    /// ensure that place have the same value from the start of the match until
-    /// an arm is selected. This prevents this code from compiling:
+    /// An immutable, aliasable borrow that is discarded after borrow-checking. Can behave either
+    /// like a normal shared borrow or like a special shallow borrow (see [`FakeBorrowKind`]).
+    ///
+    /// This is used when lowering index expressions and matches. This is used to prevent code like
+    /// the following from compiling:
+    /// ```compile_fail,E0506
+    /// let mut x = vec![vec![0, 1]];
```

Comment:
Fake borrows are only needed for built-in indexing. Something like this (note that it give a different error message):
```rust
    let mut x: &[_] = &[[0, 1]];
    let y: &[_] = &[];
    let _ = x[0][{x = y; 1}];
```

Reply:
Oh of course, this only applies to non-overloaded indexing